Add `TypeName` APIs to simplify metadata lookup.

teo-tsirpanis opened the pull request:
Fixes #101774.

teo-tsirpanis committed:
Add `TypeName.Namespace` and tests.

teo-tsirpanis committed:
Add `TypeName.Unescape`.

teo-tsirpanis committed:
Fix infinite loops.

teo-tsirpanis committed:
Simplify loop.

teo-tsirpanis committed:
Update reference assembly.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -135,6 +163,50 @@ static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
+                return input;
+            }
+
+            ValueStringBuilder builder = new(stackalloc char[128]);
```
stackalloc is relatively expensive when it is not reachable. It should be in the helper method too.

teo-tsirpanis replied:
Done.

jkotas commented:
Could you please delete https://github.com/dotnet/runtime/blob/29013d8ae50f5bc35427a9155234ccebfa5e227c/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L1492-L1494 and replace the calls to it with this API?

(Other uses of TypeNameHelpers.cs should be delete too once these APIs are available in the LKG runtime.)

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -135,6 +163,50 @@ static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
+                    }
+                    builder.Append(input.Slice(0, indexOfEscapeChar));
+                    int indexOfNextChar = indexOfEscapeChar + 1;
+                    if (indexOfNextChar < input.Length && input[indexOfNextChar] is char c && NeedsEscaping(c))
```
You should not need to call NeedsEscaping here. Take a look how the existing Unescape helper in TypeNameHelpers.cs is implemented.

jkotas replied:
Given that this is very rare case, it is better to keep the implementation as simple as possible - the existing Unescape helper in TypeNameHelpers.cs that processes one character at a time looks simpler.

teo-tsirpanis replied:
Copied implementation from `TypeNameHelpers.cs`, removed it and updated usages.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -401,6 +431,22 @@ public static bool TryParse(ReadOnlySpan<char> typeName, [NotNullWhen(true)] out
+        /// </summary>
+        /// <param name="name">The input string containing the name to convert.</param>
+        /// <returns>A string of characters with any escaped characters converted to their unescaped form.</returns>
+        /// <remarks>The unescaped string can be used for looking up the type name or namespace in metadata.</remarks>
```
Also mention that the API does not validate that the escape sequences are only used for characters that require escaping?

teo-tsirpanis replied:
Updated.

jkotas commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs:
```diff
@@ -13,13 +13,14 @@ namespace System.Reflection.Metadata.Tests
     public class TypeNameTests
```
Is there test coverage for parsing a nested types with namespace e.g. `TypeName.Parse("a.b+c.d")`?

What is `TypeName.Parse("a.b+c.d")` going to return?

teo-tsirpanis replied:
Updated to return `a` in your example.

jkotas replied:
> Passing System.Collections.Generic.Dictionary`2+MyRandomNamespace.Enumerator to Type.GetType returns the expected result

.NET Framework throws exception in this case. This looks like a bug in .NET Core - regression introduced by rewrite of the type name parser in C#. I do not think that the parser should be silently dropping any significant part of the input to the floor.

> Updated to return a in your example.

I think that the behavior for my example should be:
- Namespace returns "c", Name returns "d"
or
- Namespace returns empty string, Name returns "c.d"

teo-tsirpanis replied:
I think the former option (Namespace returns `c`, Name does not change) makes more sense and will update the PR.

There's also the question of what `a.b+c.d+e` will return, and I think the answer should be `c` as well.

teo-tsirpanis replied:
Done.

jkotas replied:
> a.b+c.d+e will return, and I think the answer should be c as well.

If it returns `c`, how does one distinguish between `a.b+c.d+e` and `a.b+c.d+c.e`?

I have been looking at handling of namespaces of nested types throughout the system. It seems that the system as a whole assumes that the nested types do not have namespaces. Would it best to avoid all handling of namespaces for nested types and treat the nested type name as a name without namespace even when it has a dot in it?

`a.b+c.d` -> Name: "c.d", Namespace: throws InvalidOperationException
`a.b+c.d+e` -> Name: "e", Namespace: throw InvalidOperationException

Alternatively, type.Namespace can return the outer most namespace instead of throwing to match reflection. However, it seems to be a pit of failure. Namespaces for nested types do not make sense. We have opted to throw InvalidOperationException in situations like this in the Typename APIs. For example, Typename.GetElementType throws InvalidOperationException, but the matching reflection API returns null.)

teo-tsirpanis replied:
Updated `GetNamespace` to throw.

adamsitnik replied:
FWIW I am supportive of the "throw" decision for nested types. And I wish things like `a.b+c.d` were illegal ;)

teo-tsirpanis committed:
Address PR feedback around `Unescape`.

teo-tsirpanis committed:
Remove file with duplicate `Unescape` method.

teo-tsirpanis committed:
Fix tests.

teo-tsirpanis committed:
Reduce allocations when calling `Namespace` across a type name hierachy.

teo-tsirpanis committed:
Fix nested types with namespaces.

teo-tsirpanis committed:
Add tests for `Unescape`.

teo-tsirpanis committed:
Fix compile errors.

teo-tsirpanis committed:
Restore `TypeNameHelpers` and use it in all places except CoreLib.

teo-tsirpanis committed:
Merge branch 'main' into typename-namespace-unescape

teo-tsirpanis committed:
Do not omit escape character at the end when unescaping.

teo-tsirpanis committed:
Return the namespace of the innermost nested type that has one.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -102,6 +102,54 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
+            {
+                ReadOnlySpan<char> nestedFullName = fullName.Slice(nestingBoundaryPos + 1);
+                offset = LastIndexOfUnescaped(nestedFullName, '.');
+                if (offset > 0)
```
Should these comparisons use `>=` instead of `>`?

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -135,6 +183,47 @@ static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
+                    }
+                }
+
+                string result = sb.ToString();
+                sb.Dispose();
+                return result;
```
```suggestion
                return sb.ToString();
```
ValueStringBuilder.ToString calls Dispose

teo-tsirpanis replied:
Done.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -401,6 +436,25 @@ public static bool TryParse(ReadOnlySpan<char> typeName, [NotNullWhen(true)] out
+        {
+            if (name is null)
+            {
+                TypeNameParserHelpers.ThrowArgumentNullException(nameof(name));
```
```suggestion
                Throw.ArgumentNull(nameof(name));
```
Can we use the pre-existing helper from System.Reflection.Metadata?

teo-tsirpanis replied:
I'm not sure, the `Throw` file is not available in Corelib. I can add an `#if SYSTEM_REFLECTION_METADATA` and just `throw` otherwise, if you prefer that.

teo-tsirpanis replied:
Or I can make `TypeName.Unescape` exclusive to SRM, and use `TypeNameParserHelpers.Unescape` in Corelib.

jkotas replied:
Ah, I forgot about sharing with CoreLib.

teo-tsirpanis committed:
Update `TypeName.Name` to return the whole name of nested types.

teo-tsirpanis committed:
Remove unnecessary `ValueStringBuilder.Dispose`.

teo-tsirpanis committed:
Update `GetNamespace` to fail if a nested type has a namespace.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -102,29 +102,54 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
-            // The two-value form of MemoryExtensions.LastIndexOfAny does not suffer
-            // from the behavior mentioned in the comment at the top of GetFullTypeNameLength.
-            // It always takes O(m * i) worst-case time and is safe to use here.
+            // Detect if a nested type name has a namespace (e.g. "a.b+c") and throw if it does.
```
I meant that the Namespace property should unconditionally throw for nested types, irrespective of whether the nested type name has a dot in it.

I am sorry that I was not clear about the suggested behavior.

teo-tsirpanis replied:
Updated to always throw in the `Namespace` property, and to just slice after the last dot in `GetNamespace`.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -102,29 +102,54 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
             {
-                offset = GetUnescapedOffset(fullName, startIndex: offset);
+                // Look for dots only if the type is not nested.
+                offset = LastIndexOfUnescaped(fullName, '.');
```
dots are never escaped. Escaped dot should have been rejected by the parser earlier. I think this can use regular last index of.

teo-tsirpanis replied:
Done. This removes a couple test cases.

jkotas commented:
There are some failing tests that needs updating. This change is fixing .NET Framework -> .NET Core regression in type name parsing. This change is going back to .NET Framework behavior that made more sense.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -102,29 +102,54 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
-            int offset = fullName.LastIndexOfAny('.', '+');
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        {
+            // If the type is nested, return the whole name after the plus sign.
```
This will be do more work than before for non-nested types - it will need to go through the full name instead of stopping at the first dot. Would it be better if the caller passed in the information about whether the type is nested or not to avoid that?

teo-tsirpanis replied:
Added an `IsNested` parameter and updated uses.

teo-tsirpanis committed:
Remove support for nested types and escaped dots in namespaces.

teo-tsirpanis committed:
Remove support for escaped dots in `GetName`, and optimize it if the type is not nested.

teo-tsirpanis commented:
> There are some failing tests that needs updating. This change is fixing .NET Framework -> .NET Core regression in type name parsing. This change is going back to .NET Framework behavior that made more sense.

We fail to parse ``.Outside`1+.Inside`1``, which looks related to this comment. I don't understand what do I need to change. https://github.com/dotnet/runtime/blob/843f7985e470f5526799abacb81cbb245c7dec70/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs#L253-L256

jkotas commented:
> We fail to parse .Outside`1+.Inside`1, which looks related to this comment. I don't understand what do I need to change.

I think we should match the .NET Framework behavior for these corner cases. For example:
```
Type.GetType("A+.B", throwOnError: true);
// Type.GetType("A+Whatever.B", throwOnError: true); - this should fail too, just like it failed in .NET Framework

public class A
{
    public class B
    {
    }
}
```
failed in .NET Framework. It "works" in .NET Core today, but I do not think that it should. I would make it fail again and update the tests as necessary.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -102,29 +102,41 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
+                    if (str[offset] == c)
                     {
-                        if (offset == 0 || fullName[offset - 1] != EscapeCharacter)
+                        if (offset == 0 || str[offset - 1] != EscapeCharacter)
```
This does not look correct for a type name like: `NameEndingInBackSlash\\+NestedName`

Instead of looking for `+` and figuring out whether it is escaped, would it be better to just get the length of the declaring type typename and slice it from there?

teo-tsirpanis replied:
Good idea, done and added your test case.

jkotas commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs:
```diff
@@ -87,6 +96,16 @@ public void ParserIsNotEnforcingRuntimeSpecificRules(string input)
+        [InlineData("System.Int32[]", "System.Int32[]")]
+        [InlineData("System.Int32\\[\\]", "System.Int32[]")]
+        [InlineData("System.Int32\\", "System.Int32\\")]
+        [InlineData("System.Int32\\\\[]", "System.Int32\\[]")]
```
```suggestion
        [InlineData("System.Int32\\\\[]", "System.Int32\\[]")]
        [InlineData("System\\.Int32", "System.Int32")]
```
This validates that the API does not check that the escape sequence is valid.

teo-tsirpanis replied:
Added.

teo-tsirpanis committed:
Update tests.

teo-tsirpanis committed:
Simplify `Name` to avoid linear search of the full name in nested types.

teo-tsirpanis commented:
Updated tests to account for ``.Outside`1+.Inside`1`` failing to parse.

jkotas commented:
Could you please take a look at the Mono test failures?

jkotas reviewed (approved):
Looks good to me. @adamsitnik Could you please review this change as well? We are changing the type name parsing behavior for nested types with its own namespaces to fix .NET Framework -> .NET Core type name parsing regression and to better match reflection behavior.

teo-tsirpanis commented:
```
    System.Tests.TypeTests.GetTypeByName_Invalid(typeName: ".Outside`1+.Inside`1", expectedException: typeof(System.TypeLoadException), alwaysThrowsException: False) [FAIL]
      Assert.Null() Failure: Value is not null
      Expected: null
      Actual:   typeof(Outside<,>)
      Stack Trace:
        /_/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs(547,0): at System.Tests.TypeTests.GetTypeByName_Invalid(String typeName, Type expectedException, Boolean alwaysThrowsException)
           at System.Object.InvokeStub_TypeTests.GetTypeByName_Invalid(Object , Span`1 )
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(136,0): at System.Reflection.MethodBaseInvoker.InvokeWithFewArgs(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

teo-tsirpanis commented:
Mono thinks `GetNestedType` accepts a full name, while it does not [per the documentation](https://learn.microsoft.com/en-us/dotnet/api/system.type.getnestedtype?view=net-9.0#system-type-getnestedtype(system-string)). https://github.com/dotnet/runtime/blob/8eab62c365de53e0a341fb1000249d338c203354/src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs#L1011-L1018

jkotas commented:
> Mono thinks GetNestedType accepts a full name, while it does not [per the documentation](https://learn.microsoft.com/en-us/dotnet/api/system.type.getnestedtype?view=net-9.0#system-type-getnestedtype(system-string)).

Looks like something that should be cleaned up. Do you have a good idea what it needs to be done?

teo-tsirpanis committed:
[mono] Do not treat nested type names as full names.

teo-tsirpanis committed:
Fix compile errors.

teo-tsirpanis commented:
I tried something. Let's see.

jkotas commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -1008,22 +1004,19 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
 
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicNestedTypes | DynamicallyAccessedMemberTypes.NonPublicNestedTypes)]
-        public override Type? GetNestedType(string fullname, BindingFlags bindingAttr)
+        public override Type? GetNestedType([MaybeNull] string name, BindingFlags bindingAttr)
```
These are public overridden methods. Does `[MaybeNull]` impact how nullability of these methods is going to be seen via reflection?

teo-tsirpanis replied:
[Yes](https://github.com/dotnet/runtime/blob/d4c7d865558316da4c135f519eb3dfc25b8a4367/src/libraries/System.Private.CoreLib/src/System/Reflection/NullabilityInfoContext.cs#L156), but the fact that `name` might be reassigned to `null` is not meaningful because the parameter is passed by value.

jkotas replied:
I would just copy it into a local variable.

teo-tsirpanis replied:
Reverted changes and kept status quo.

adamsitnik commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs:
```diff
@@ -87,6 +96,42 @@ public void ParserIsNotEnforcingRuntimeSpecificRules(string input)
+                }
+                parsed = parsed.DeclaringType;
+                i--;
+            }
```
nit: it would be nice to ensure that the loop checks all the expected names

```suggestion
            }
            Assert.Equal(0, i);
```

teo-tsirpanis replied:
Done.

adamsitnik commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Type/TypeTests.cs:
```diff
@@ -539,6 +538,7 @@ public static IEnumerable<object[]> GetTypeByName_InvalidElementType()
         [InlineData("Outside`1[System.Boolean, System.Int32]", typeof(ArgumentException), true)]
         [InlineData(".System.Int32", typeof(TypeLoadException), false)]
         [InlineData("..Outside`1", typeof(TypeLoadException), false)]
+        [InlineData(".Outside`1+.Inside`1", typeof(TypeLoadException), false)]
```
Should this be documented as a breaking change?

adamsitnik commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -217,6 +217,7 @@ public string FullName
         /// This is because determining whether a type truly is a generic type requires loading the type
         /// and performing a runtime check.</para>
         /// </remarks>
+        [MemberNotNullWhen(false, nameof(_elementOrGenericType))]
```
nice, I had no idea that such an attribute existed 👍

adamsitnik commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -262,28 +264,82 @@ public string Name
                     if (IsConstructedGenericType)
                     {
-                        _name = TypeNameParserHelpers.GetName(GetGenericTypeDefinition().FullName.AsSpan()).ToString();
+                        _name = GetGenericTypeDefinition().Name;
```
nice simplification 👍

adamsitnik reviewed (approved):
LGTM, thank you for your contribution @teo-tsirpanis !

And big thanks to @jkotas for the review!

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -102,36 +102,48 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
-            // It always takes O(m * i) worst-case time and is safe to use here.
-
-            int offset = fullName.LastIndexOfAny('.', '+');
+            int offset = fullName.LastIndexOf('.');
```
> According to Copilot, IL allows for creating a type with a name that contains a dot that is not treated as a namespace separator.

That does not sound correct to me. The namespace/name splitting of the full name is really just a metadata encoding optimization to avoid storing namespaces repeatedly. The ECMA-335 says: "While some programming languages introduce the concept of a namespace, the only support in the CLI
for this concept is as a metadata encoding technique. Type names are always specified by their full
name relative to the assembly in which they are defined.".

Unfortunately, ECMA-335 does not specify the namespace/name splitting algorithm. It says that you can use any dot for the split, but that makes things complicated, so it is not used in practice.

The namespace/name splitting algorithm that has been used in .NET runtime since forever is:

> The last dot in the name of non-nested type is always treated as a namespace separator, except when there is a dot immediately preceding the last dot - the second to last dot is treated as a namespace separator in that case instead. 

If you use the reflection emit test above to explore various cases, you should see following behavior:
`Namespace..Name` -> namespace=`Namespace`, name = `.Name`
`Namespace...Name` -> namespace=`Namespace.`, name = `.Name`
`Namespace..Name.` -> namespace=`Namespace..Name`, name = ``

@teo-tsirpanis Could you please add coverage for these cases?

I think we should fix the type name parser implementation to follow it to avoid inconsistencies. This algorithm is implemented by https://github.com/dotnet/runtime/blob/d20c35e05c16dd2ec9cddd4259738cc02228d425/src/libraries/Common/src/System/Reflection/Metadata/TypeNameHelpers.cs#L56-L77 that we hope to replace by the new APIs.

teo-tsirpanis replied:
Done.

teo-tsirpanis committed:
Update algorithm to find namespace delimiter.

teo-tsirpanis committed:
Revert nullable annotations changes.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -262,28 +264,89 @@ public string Name
+        {
+            get
+            {
+                if (IsNested)
```
Should this check be done on rootTypeName?

It does not make sense for "MyNamespace.A+B[]" to return `MyNamespace` when "MyNamespace.A+B" fails with InvalidOperationException.

teo-tsirpanis replied:
Would it make sense to throw in arrays, pointers and refs, like in nested types?

jkotas replied:
I do not have a strong opinion. 

If I had to choose, I would throw for nested types only. I think that the pit of failure that we want to protect from is limited to nested types. If you have `MyNamespace.A[]`, it makes sense that `Name` returns `A[]` and `Namespace` returns `MyNamespace`.

teo-tsirpanis replied:
Makes sense, done.

jkotas commented:
One test is still failing on Mono

teo-tsirpanis commented:
> One test is still failing on Mono

I took another look at the sources and have no idea on why it fails. Might try debugging it, once I manage to build Mono on Windows.

teo-tsirpanis committed:
Redirect all `Type.GetType` overloads in Mono through `TypeNameResolver`.

teo-tsirpanis committed:
Update tests to account for Mono using the managed type parser in `Type.GetType`.

teo-tsirpanis commented:
One Mono test fails because the thrown `TypeLoadException` does not have its `TypeName` set.

The other two are more weird:

```
System.Reflection.Tests.GetTypeTests.GetTypeTest(typeName: "System.Reflection.Tests.MyNamespace1.MyNamespace2."···, expectedResult: typeof(System.Reflection.Tests.MyNamespace1.MyNamespace2.MyClass3+iNner)) [FAIL]
  System.Reflection.AmbiguousMatchException : Ambiguous match found for 'System.Reflection.Tests.MyNamespace1.MyNamespace2.MyClass3 System.Reflection.Tests.MyNamespace1.MyNamespace2.MyClass3+Inner'.
  Stack Trace:
    /_/src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs(1024,0): at System.RuntimeType.GetNestedType(String name, BindingFlags bindingAttr)
    /_/src/mono/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.Mono.cs(143,0): at System.Reflection.TypeNameResolver.GetType(String escapedTypeName, ReadOnlySpan`1 nestedTypeNames, TypeName parsedName)
    /_/src/libraries/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.cs(87,0): at System.Reflection.TypeNameResolver.GetSimpleType(TypeName typeName)
    /_/src/libraries/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.cs(36,0): at System.Reflection.TypeNameResolver.Resolve(TypeName typeName)
    /_/src/mono/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.Mono.cs(49,0): at System.Reflection.TypeNameResolver.GetType(String typeName, Func`2 assemblyResolver, Func`4 typeResolver, Boolean throwOnError, Boolean ignoreCase, StackCrawlMark& stackMark)
    /_/src/mono/System.Private.CoreLib/src/System/Type.Mono.cs(79,0): at System.Type.GetType(String typeName, Func`2 assemblyResolver, Func`4 typeResolver, Boolean throwOnError, Boolean ignoreCase, StackCrawlMark& stackMark)
    /_/src/mono/System.Private.CoreLib/src/System/Type.Mono.cs(33,0): at System.Type.GetType(String typeName, Boolean throwOnError, Boolean ignoreCase)
    /_/src/libraries/System.Runtime/tests/System.Reflection.Tests/GetTypeTests.cs(180,0): at System.Reflection.Tests.GetTypeTests.GetTypeTest(String typeName, Type expectedResult)
        at System.Object.InvokeStub_GetTypeTests.GetTypeTest(Object , Span`1 )
    /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(136,0): at System.Reflection.MethodBaseInvoker.InvokeWithFewArgs(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

jkotas commented:
> Ambiguous match found

I think that's because the managed type name resolver does not have this: https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Reflection/TypeNameResolver.CoreCLR.cs#L239

teo-tsirpanis committed:
Disallow getting the namespace of non-simple nested type names.

teo-tsirpanis committed:
Ignore ambiguous match exceptions in non-extensible `Type.GetType` overloads.

This matches the behavior of CoreCLR and Native AOT.

teo-tsirpanis committed:
Set `TypeLoadException.TypeName` in Mono as well.

The property became init-only (internally) in order to avoid conflicts with runtime-specific constructor overloads.

teo-tsirpanis committed:
Update src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

teo-tsirpanis committed:
Merge branch 'main' into typename-namespace-unescape

jkotas commented on src/libraries/System.Private.CoreLib/src/System/TypeLoadException.cs:
```diff
@@ -29,15 +29,6 @@ public TypeLoadException(string? message, Exception? inner)
         }
 
-#if !MONO
-        internal TypeLoadException(string message, string typeName)
```
This constructor is used internally by CoreCLR to construct TypeLoadExceptions. If you delete it, you will have to redo how type load exceptions are created internally by CoreCLR.

It would be much less invasive to keep it, enable it for Mono and fix this exclusively on the Mono side.

jkotas replied:
Disabling the failing test on Mono would be fine too.

teo-tsirpanis replied:
> This constructor is used internally by CoreCLR to construct TypeLoadExceptions. If you delete it, you will have to redo how type load exceptions are created internally by CoreCLR.

I think I redid it? CI is mostly green apart from unrelated failures, and this overload was added three weeks ago in #111374 and there weren't any uses from C++ there.

jkotas replied:
I was worried about the places where the VM creates the exception using "reflection" like: https://github.com/dotnet/runtime/blob/16782a4481f389e68314b16ea628c5af80f22783/src/coreclr/vm/clrex.cpp#L1471-L1472 . A good point that this constructor was added just recently, so it is not used like that. 

Still, I am not sure whether the init-only is an improvement. If we were to introduce a public API to allow creating TypeLoadException with type name, I think we would add a constructor that takes the additional argument for consistency with how all existing BCL exception constructors look.

teo-tsirpanis replied:
Yes, a `(string message, string typeName)` constructor would be the ideal API shape but that would require changes in Mono's C code, which I would prefer to not expand this PR's scope with.

teo-tsirpanis replied:
I thought of a (likely hacky) way that both simulates a `(message, typeName)` overload, and avoids changes to the Mono runtime. Let me know what you think.

jkotas replied:
I am ok with it. I left a few comments.

teo-tsirpanis committed:
Use alternative strategy to provide a `(message, typeName)` overload in Mono, while still avoiding changes to the Mono runtime code.

jkotas commented on src/mono/System.Private.CoreLib/src/System/TypeLoadException.Mono.cs:
```diff
@@ -13,6 +13,18 @@ internal TypeLoadException(string className, string assemblyName)
             _assemblyName = assemblyName;
```
Can me make the `string className, string assemblyName` constructor private, so that it is hard to call it by mistake from random places? (If it is private, it can be only called from within the TypeLoadException implementation.)

teo-tsirpanis replied:
Done, and suppressed the suggestion to remove the constructor.

jkotas commented on src/mono/System.Private.CoreLib/src/System/TypeLoadException.Mono.cs:
```diff
@@ -13,6 +13,18 @@ internal TypeLoadException(string className, string assemblyName)
+        // we add a dummy parameter with a default value to minimize native code changes.
+        // In order to use this overload, callers should either pass three arguments, or specify
+        // one of the parameters by name.
+        internal TypeLoadException(string message, string typeName, bool useMessageTypeNameOverload = true)
+            : base(message)
+        {
+            _ = useMessageTypeNameOverload;
```
```suggestion
        internal TypeLoadException(string message, string typeName, bool _ = true)
            : base(message)
        {
```

teo-tsirpanis replied:
Done.

jkotas commented on src/mono/System.Private.CoreLib/src/System/TypeLoadException.Mono.cs:
```diff
@@ -13,6 +13,18 @@ internal TypeLoadException(string className, string assemblyName)
             _assemblyName = assemblyName;
         }
 
+        // Because the Mono runtime has a dependency to a (string, string) constructor overload,
+        // we add a dummy parameter with a default value to minimize native code changes.
+        // In order to use this overload, callers should either pass three arguments, or specify
+        // one of the parameters by name.
```
```suggestion
        // Because the Mono runtime has a dependency on a (string, string) constructors of exception types,
        // we add a dummy parameter with a default value.
        // In order to use this overload, callers should specify
        // the typeName parameter by name.
```

teo-tsirpanis replied:
Done.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs:
```diff
@@ -102,36 +102,56 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
             {
-                int offset = startIndex;
-                for (; offset >= 0; offset--)
+                // this code path is executed very rarely (IL Emit or pure IL with chars not allowed in C# or F#)
```
```suggestion
                // This code path is executed very rarely (IL Emit or pure IL with chars not allowed in C# or F#)
```

teo-tsirpanis replied:
Done.

teo-tsirpanis committed:
Address PR feedback.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs:
```diff
@@ -262,28 +264,89 @@ public string Name
+                        }
+                        else if (TypeNameParserHelpers.IndexOfNamespaceDelimiter(name) is int idx && idx >= 0)
+                        {
+                            // If the type is not nested, find the namespace delimiter in the full name and and return the substring after it.
```
```suggestion
                            // If the type is not nested, find the namespace delimiter in the full name and return the substring after it.
```

jkotas committed:
Update src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

jkotas commented:
I have opened https://github.com/dotnet/runtime/issues/112376 as a follow up to use these APIs in more places.

jkotas commented:
> It's strange, this regression is WASM specific

I expect that the regression is Mono interpreter specific.   

This change https://github.com/dotnet/runtime/pull/111598/files#diff-02398900754074bfbfe1f634cd44c490e92f433d7cdd8eb217effde64b3581baR49 switched `Type.GetType(string)` to call common type parsing code to avoid bugs and behavior differences in Mono C type parsing code. The common type parsing code is written in C# that is slow when executed with Mono interpreter.

If we care about this regression, we can revert the change in Type.Mono.cs to fix it. It will re-introduce bugs and behavior differences on Mono that will need to be dealt with.
